Spigot-Server-Patches/0444-Allow-chunks-to-use-the-persistent-data-API.patch

```diff
@@ -0,0 +1,83 @@
+                 loadEntities(nbttagcompound1, chunk);
++                readBukkitValues(nbttagcompound1, chunk); // Paper - Load persistent data from NBT
+             });
+         } else {
+             ProtoChunk protochunk = new ProtoChunk(chunkcoordintpair, chunkconverter, achunksection, protochunkticklist, protochunkticklist1, worldserver); // Paper - Anti-Xray
+@@ -455,6 +456,8 @@ public class ChunkRegionLoader {
+                 }
+             }
+ 
++            storeBukkitValues(nbttagcompound1, chunk); // Paper - Store persistent data to NBT
++
+             // Paper start - move entities to the correct chunk
+             for (Entity entity : toUpdate) {
+                 worldserver.entityJoinedWorld(entity);
+@@ -583,6 +586,22 @@ public class ChunkRegionLoader {
+ 
+     }
+ 
++    // Paper start - Handle persistent data reading and writing
++    public static void readBukkitValues(NBTTagCompound nbtTagCompound, Chunk chunk) {
++        NBTTagCompound persistentBase = nbtTagCompound.getCompound("BukkitValues");
++        if (persistentBase != null) {
++            ((org.bukkit.craftbukkit.CraftChunk) chunk.getBukkitChunk()).persistentDataContainer.putAll(persistentBase);
++        }
++    }
++
++    public static void storeBukkitValues(NBTTagCompound nbtTagCompound, Chunk chunk) {
++        org.bukkit.craftbukkit.CraftChunk craftChunk = ((org.bukkit.craftbukkit.CraftChunk) chunk.getBukkitChunk());
++        if (!craftChunk.persistentDataContainer.isEmpty()) {
++            nbtTagCompound.set("BukkitValues", craftChunk.persistentDataContainer.toTagCompound());
++        }
++    }
++    // Paper end
++
+     private static NBTTagCompound a(ChunkCoordIntPair chunkcoordintpair, Map<String, StructureStart> map, Map<String, LongSet> map1) {
+         NBTTagCompound nbttagcompound = new NBTTagCompound();
+         NBTTagCompound nbttagcompound1 = new NBTTagCompound();
+diff --git a/src/main/java/org/bukkit/craftbukkit/CraftChunk.java b/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
+index 9cff8b88..49e63317 100644
+--- a/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
++++ b/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
+@@ -42,6 +42,11 @@ public class CraftChunk implements Chunk {
+     private static final DataPaletteBlock<IBlockData> emptyBlockIDs = new ChunkSection(0).getBlocks();
+     private static final byte[] emptyLight = new byte[2048];
+ 
++    // Paper start
++    private static final org.bukkit.craftbukkit.persistence.CraftPersistentDataTypeRegistry DATA_TYPE_REGISTRY = new org.bukkit.craftbukkit.persistence.CraftPersistentDataTypeRegistry();
++    public org.bukkit.craftbukkit.persistence.CraftPersistentDataContainer persistentDataContainer = new org.bukkit.craftbukkit.persistence.CraftPersistentDataContainer(DATA_TYPE_REGISTRY);
```

Comment:
The persistent data container needs to be on the Chunk object (should be private final too). Don't need to deal with plugins that leak CraftChunk instances to also leak the persistent data.

Reply:
This, actually, I forgot that people hold onto CraftChunk's \o/